### ISSUE TITLE
createTheme to css variable script

### ISSUE THREAD

**Issue by jeremy-coleman** [OfficeDev/office-ui-fabric-react#8970]
copy and paste this into vscode and exec with node to see output,

in order to to use it in the browser, 
 call createTheme with convertObjectToCssVariables to remap the vars ie: themeDarker: 'var(--themeDarker)'. the second set of json is used as an input parameter for document.querySelector(#anything).style.setProperty(ThoseVars)). if you forEach or map all of them targeting whatever selector, it's the same result as a adding another customizer , except you're only creating ~20 dom props and won't over-write anything above it. you can also reset it by just calling the defaults on whatever child , with pretty much 0 overhead also no react props are needed to delegate the theme what so ever. pretty easy win in ~50 lines of code for ie11+. the applyTheme fn below applies to body, but if you want to make a <Customizer/> equivalent, just make a div id="some-brand-name" and querySelect it instead of body. you can also select individual fabric components by using querySelctorAll and searching for some partial match on the classNames. you could use a mix of the two strategies to do something like, apply the Ms Teams theme only to fabric button components only at a certain depth. without context or customizer and without re-creating a billion copies of fabric in memory

```js
// When adding or removing a color, make sure you keep this consistent with IColorClassNames by adding the color variants.
const DefaultPalette = {
  themeDarker: '#004578',
  themeDark: '#005a9e',
  themeDarkAlt: '#106ebe',
  themePrimary: '#0078d4',
  themeSecondary: '#2b88d8',
  themeTertiary: '#71afe5',
  themeLight: '#c7e0f4',
  themeLighter: '#deecf9',
  themeLighterAlt: '#eff6fc',
  black: '#000000',
  blackTranslucent40: 'rgba(0,0,0,.4)',
  neutralDark: '#212121',
  neutralPrimary: '#333333',
  neutralPrimaryAlt: '#3c3c3c',
  neutralSecondary: '#666666',
  neutralSecondaryAlt: '#767676',
  neutralTertiary: '#a6a6a6',
  neutralTertiaryAlt: '#c8c8c8',
  neutralQuaternary: '#d0d0d0',
  neutralQuaternaryAlt: '#dadada',
  neutralLight: '#eaeaea',
  neutralLighter: '#f4f4f4',
  neutralLighterAlt: '#f8f8f8',
  accent: '#0078d4',
  white: '#ffffff',
  whiteTranslucent40: 'rgba(255,255,255,.4)',
  yellow: '#ffb900',
  yellowLight: '#fff100',
  orange: '#d83b01',
  orangeLight: '#ea4300',
  orangeLighter: '#ff8c00',
  redDark: '#a80000',
  red: '#e81123',
  magentaDark: '#5c005c',
  magenta: '#b4009e',
  magentaLight: '#e3008c',
  purpleDark: '#32145a',
  purple: '#5c2d91',
  purpleLight: '#b4a0ff',
  blueDark: '#002050',
  blueMid: '#00188f',
  blue: '#0078d4',
  blueLight: '#00bcf2',
  tealDark: '#004b50',
  teal: '#008272',
  tealLight: '#00b294',
  greenDark: '#004b1c',
  green: '#107c10',
  greenLight: '#bad80a'
};

  function _createStyleElement() {
    const styleElement = document.createElement('style');

    styleElement.setAttribute('data-merge-styles', 'true');
    styleElement.type = 'text/css';

    if (this._lastStyleElement && this._lastStyleElement.nextElementSibling) {
      document.head.insertBefore(styleElement, this._lastStyleElement.nextElementSibling);
    } else {
      document.head.appendChild(styleElement);
    }
    this._lastStyleElement = styleElement;

    return styleElement;
  }




 function nodeTheme(themeObject) {
  Object.entries(themeObject).forEach(_ => console.log(`"--${_[0]}", ${_[1]}`));
}

function update(current, updater) {
  if (
    Object.prototype.toString.call(current) === "[object Object]" ||
    Array.isArray(current)
  ) {
    var draft = JSON.parse(JSON.stringify(current));
    updater(draft);
    return draft;
  }
}

function convertObjectToCssVariables(themeObject){
  Object.entries(themeObject).forEach(_ => console.log(`${_[0]}`))
}

function convertObjectToCssVariables2(themeObject){
  update(DefaultPalette, draft => Object.keys(draft), )
}

function convertObjectToCssVariables3(themeObject){
  let a = []
  let b = {}
  
  let x = Array.from(Object.keys(themeObject))
  let y = x.map(_ => `--${_}`)
  
  for (k in themeObject){
    a.push([k, `var(--${k})`])
    //console.log()
  }
  let c = new Map(a)
  console.log(c)
}


function convertObjectToCssVariables(themeObject){
  let a = Object.assign({}, themeObject)
  let x = Object.keys(themeObject).map(_ => a[_] = `var(--${_})`)
  console.log(a)
}


function convertObjectToCssVariables4(themeObject){
  let _keys = Object.entries(themeObject).map(_ => `${_[0]}`)
  let cssVars = _keys.map(_ => `var(--${_})`)
  let x = cssVars
  console.log(_keys, cssVars)
}


function applyTheme(themeObject) {
  Object.entries(themeObject).forEach(_ => document.body.style.setProperty(`"--${_[0]}", ${_[1]}`));
}


convertObjectToCssVariables(DefaultPalette)
nodeTheme(DefaultPalette)

//@ts-ignore
//console.log(Object.entries(DefaultPalette)[0])



// thisWillPrint =
//  {
//   themeDarker: 'var(--themeDarker)',
//   themeDark: 'var(--themeDark)',
//   themeDarkAlt: 'var(--themeDarkAlt)',
//   themePrimary: 'var(--themePrimary)',
//   themeSecondary: 'var(--themeSecondary)',
//   themeTertiary: 'var(--themeTertiary)',
//   themeLight: 'var(--themeLight)',
//   themeLighter: 'var(--themeLighter)',
//   themeLighterAlt: 'var(--themeLighterAlt)',
//   black: 'var(--black)',
//   blackTranslucent40: 'var(--blackTranslucent40)',
//   neutralDark: 'var(--neutralDark)',
//   neutralPrimary: 'var(--neutralPrimary)',
//   neutralPrimaryAlt: 'var(--neutralPrimaryAlt)',
//   neutralSecondary: 'var(--neutralSecondary)',
//   neutralSecondaryAlt: 'var(--neutralSecondaryAlt)',
//   neutralTertiary: 'var(--neutralTertiary)',
//   neutralTertiaryAlt: 'var(--neutralTertiaryAlt)',
//   neutralQuaternary: 'var(--neutralQuaternary)',
//   neutralQuaternaryAlt: 'var(--neutralQuaternaryAlt)',
//   neutralLight: 'var(--neutralLight)',
//   neutralLighter: 'var(--neutralLighter)',
//   neutralLighterAlt: 'var(--neutralLighterAlt)',
//   accent: 'var(--accent)',
//   white: 'var(--white)',
//   whiteTranslucent40: 'var(--whiteTranslucent40)',
//   yellow: 'var(--yellow)',
//   yellowLight: 'var(--yellowLight)',
//   orange: 'var(--orange)',
//   orangeLight: 'var(--orangeLight)',
//   orangeLighter: 'var(--orangeLighter)',
//   redDark: 'var(--redDark)',
//   red: 'var(--red)',
//   magentaDark: 'var(--magentaDark)',
//   magenta: 'var(--magenta)',
//   magentaLight: 'var(--magentaLight)',
//   purpleDark: 'var(--purpleDark)',
//   purple: 'var(--purple)',
//   purpleLight: 'var(--purpleLight)',
//   blueDark: 'var(--blueDark)',
//   blueMid: 'var(--blueMid)',
//   blue: 'var(--blue)',
//   blueLight: 'var(--blueLight)',
//   tealDark: 'var(--tealDark)',
//   teal: 'var(--teal)',
//   tealLight: 'var(--tealLight)',
//   greenDark: 'var(--greenDark)',
//   green: 'var(--green)',
//   greenLight: 'var(--greenLight)'
// }

// "--themeDarker", #004578
// "--themeDark", #005a9e
// "--themeDarkAlt", #106ebe
// "--themePrimary", #0078d4
// "--themeSecondary", #2b88d8
// "--themeTertiary", #71afe5
// "--themeLight", #c7e0f4
// "--themeLighter", #deecf9
// "--themeLighterAlt", #eff6fc
// "--black", #000000
// "--blackTranslucent40", rgba(0,0,0,.4)
// "--neutralDark", #212121
// "--neutralPrimary", #333333
// "--neutralPrimaryAlt", #3c3c3c
// "--neutralSecondary", #666666
// "--neutralSecondaryAlt", #767676
// "--neutralTertiary", #a6a6a6
// "--neutralTertiaryAlt", #c8c8c8
// "--neutralQuaternary", #d0d0d0
// "--neutralQuaternaryAlt", #dadada
// "--neutralLight", #eaeaea
// "--neutralLighter", #f4f4f4
// "--neutralLighterAlt", #f8f8f8
// "--accent", #0078d4
// "--white", #ffffff
// "--whiteTranslucent40", rgba(255,255,255,.4)
// "--yellow", #ffb900
// "--yellowLight", #fff100
// "--orange", #d83b01
// "--orangeLight", #ea4300
// "--orangeLighter", #ff8c00
// "--redDark", #a80000
// "--red", #e81123
// "--magentaDark", #5c005c
// "--magenta", #b4009e
// "--magentaLight", #e3008c
// "--purpleDark", #32145a
// "--purple", #5c2d91
// "--purpleLight", #b4a0ff
// "--blueDark", #002050
// "--blueMid", #00188f
// "--blue", #0078d4
// "--blueLight", #00bcf2
// "--tealDark", #004b50
// "--teal", #008272
// "--tealLight", #00b294
// "--greenDark", #004b1c
// "--green", #107c10
// "--greenLight", #bad80a
```

**Comment by jdhuntington** [comment]
Thanks @jeremy-coleman -- this is a great approach; thank you!

We are actively looking at options for including CSS variables (with the obvious caveats in browser support). An approach that lets someone selectively opt-in with class names rather than diving all the way in is a great idea too. 

To discuss specifics, I expect we will be able to focus on this more in the near future after we've released the 7.0 version.

**Comment by jeremy-coleman** [comment]
@dearzing @jdhuntington
Np, ive been using this approach for a few months, one adjustment im planning to make is use ts vars for the var strings for better editor feedback. Using the new ‘as const’ keyword in ts 3.4 makes it super easy
